astropy/utils/data.py

```diff
@@ -1327,43 +1406,303 @@ def _get_download_cache_locs():
+            warn(CacheMissingWarning(msg + e.__class__.__name__ + estr))
+            yield None, nothing
+            return
+    if write:
+        with _cache_lock(need_write=True), \
+                shelve.open(urlmapfn, flag="c") as url2hash:
+            yield dldir, url2hash
     else:
-        msg = 'Error releasing lock. "{0}" either does not exist or is not ' +\
-              'a directory.'
-        raise RuntimeError(msg.format(lockdir))
+        try:
+            with _cache_lock(), shelve.open(urlmapfn, flag="r") as url2hash:
+                # Copy so we can release the lock.
+                d = ReadOnlyDict(url2hash.items())
+        except dbm.error:
+            # Might be a "file not found" - that is, an un-initialized cache,
+            # might be something serious, no way to tell as shelve just gives
+            # you a plain dbm.error
+            # Also the file doesn't have a platform-independent name, so good
+            # luck diagnosing the problem if it is one.
+            d = nothing
+        yield dldir, d
+
+
+class CacheDamaged(ValueError):
+    """Record the URL or file that was a problem.
+
+    Using clear_download_cache on the .bad_file or .bad_url attribute,
+    whichever is not None, should resolve this particular problem.
+    """
+    def __init__(self, *args, bad_url=None, bad_file=None, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.bad_url = bad_url
+        self.bad_file = bad_file
+
+
+def check_download_cache(check_hashes=False):
+    """Do a consistency check on the cache
+
+    Because the cache is shared by all versions of astropy in all virtualenvs
+    run by your user, possibly concurrently, it could accumulate problems.
+    This could lead to hard-to-debug problems or wasted space. This function
+    detects a number of incorrect conditions, including nonexistent files that
+    are indexed, files that are indexed but in the wrong place, and, if you
+    request it, files whose content does not mash the hash that is indexed.
+
+    This funtion also returns a list of non-indexed files. A few will be
+    associated with the shelve object; their exact names depend on the backend
+    used but will probably be based on ``urlmap``. The presence of other files,
+    particularly those that are 32-character hex strings or start with those,
+    probably indicates that something has gone wrong and inaccessible files
+    have accumulated in the cache. These can be removed with
+    `clear_download_cache`, either passing the filename returned here, or
+    with no arguments to empty the entire cache and return it to a
+    reasonable, if empty, state.
+
+    Parameters
+    ----------
+    check_hashes : bool, optional
+        Whether to compute the hashes of the contents of all files in the
+        cache and compare them to the names. This can take some time if
+        the cache contains large files.
+
+    Returns
+    -------
+    strays : set of strings
```

Comment:
When a user check the cache, the user probably wants a full report, rather than just the strays? For instance, I was hoping this would give me a full list of stuff in the cache and the status of each. Something like:

* url_1 = good
* url_2 = cache damaged
* url_3 = good
* url_4 = in the wrong place
...

Reply:
In practice this is a function I call all over the test suite to make sure no operation has left the cache damaged. So in fact it probably shouldn't return anything - it raises an exception if something is wrong. The new version of the exception contains a report of all the things that are wrong. I could add a dictionary of the ok parts too if that seems useful.

The reason for having a return value is because I wanted to pick up on stray files that accumulated but we don't know what files shelve creates, so the user needs to do some fiddling to detect whether there's anything weird. It's a lot less necessary now that I catch abandoned files whose names look like hashes.

Reply:
While you might have intended it as a utility function for test suite, given that it is a public function in the module, there is no stopping users to try to use this for their own purpose. I was just thinking aloud of what I would have wanted if I stumbled upon this function in the API doc as a user.

If you really want to limit its use to test suite, we can either make this a "private" function or move it inside `tests` directory. Though, I would find this generally useful for everyone.